### DISCUSSION TITLE
How to explicitly query and map non-persisted (virtual) joined properties?

### DISCUSSION THREAD

**Discussion by squivix** [mikro-orm/mikro-orm#4545] [Q&A]
I have the following entties:

```ts
export class EntityA extends CustomBaseEntity {
    @OneToMany({entity: () => EntityB, mappedBy: (entityB: EntityB) => entityB.entityA,})
    entityBs: Collection<EntityB> = new Collection<EntityB>(this);

    @Property({type: types.integer})
    value!: number;
}
```

```ts
@Entity({})
export class EntityB extends CustomBaseEntity {
    @ManyToOne({entity: () => EntityA, inversedBy: (entityA) => entityA.entityBs})
    entityA!: EntityA;

    @Property({type: types.string})
    text!: string;

    @OneToMany({entity: () => EntityC, mappedBy: (entityC: EntityC) => entityC.entityB})
    entityCs!: EntityC;

    @Property({type: types.integer, persist: false})
    entityCsCount!: number;
}
```

```ts
@Entity({})
export class EntityC extends CustomBaseEntity {
    @ManyToOne({entity: () => EntityB, inversedBy: (entityB) => entityB.entityCs})
    entityB!: EntityB;

    @Property({type: types.boolean})
    booleanColumn!: string;
}
```

I am trying to query all `EntityA`'s with their associated `EntityB`'s each with a count of associated `EntityC`'s. Instead of using a formula (which would be slow), I am writing the query with the querybuilder:

```ts
const queryBuilder = em.createQueryBuilder(EntityA, "ea");
queryBuilder
    .select("*")
    .addSelect("COUNT(DISTINCT ec.id) AS eb__entity_cs_count")
    .leftJoinAndSelect("ea.entityBs", "eb")
    .leftJoin("eb.entityCs", "ec")
    .groupBy(["ea.id", "eb.id"])
const entityAs = await queryBuilder.getResultList();
console.log(entityAs[0].entityBs[0].entityCsCount);     //expected 1, actual undefined
```

But the value of the virtual property is not set, even though it is returned from the database and has the right alias eb__entity_cs_count.

Is there no way to explicitly fetch virtual properties and have them be set automatically? What's the alternative?

Edit:

I tracked it down to the place where the column is excluded.

https://github.com/mikro-orm/mikro-orm/blob/e5a9ed578defeecf3f6e292bd4cd0b67d0fa9340/packages/knex/src/AbstractSqlDriver.ts#L228

And then:

https://github.com/mikro-orm/mikro-orm/blob/e5a9ed578defeecf3f6e292bd4cd0b67d0fa9340/packages/core/src/platforms/Platform.ts#L428-L435

 I wonder why columns with persist: false are excluded from `mapJoinedProps`

**Comment by B4nan** [comment]
> I wonder why columns with persist: false are excluded from mapJoinedProps

Because those are virtual, meaning they don't represent a database column. You are not supposed to be using that on real database columns.

You can always try to comment out some lines and run the test suite, it will give you a better idea of why are things implemented in some way and guide you to the tests connected to something (as they will start failing).

Maybe this is something to improve, but I am quite certain it is not only about removing that `if` branch. Could you compile this into a test case?
